src/app/services/baw-api/model.service.ts

```diff
@@ -0,0 +1,143 @@
+import { HttpClient } from "@angular/common/http";
+import { Inject, Injectable, InjectionToken } from "@angular/core";
+import { Subject } from "rxjs";
+import { AppConfigService } from "../app-config/app-config.service";
+import { APIErrorDetails } from "./api.interceptor";
+import { BawApiService, Filters } from "./base-api.service";
+
+export let MOCK_CLASS_BUILDER = new InjectionToken("test.class.builder");
+
+@Injectable()
+export class ModelService<T> extends BawApiService {
+  private subjectNext(subject: Subject<T>, data: any) {
+    subject.next(this.classBuilder(data));
+    subject.complete();
+  }
+  private subjectNextList(subject: Subject<T[]>, data: any[]) {
+    subject.next(data.map(object => this.classBuilder(object)));
+    subject.complete();
+  }
+  private subjectError(subject: Subject<T | T[]>, err: APIErrorDetails) {
+    subject.error(err);
+  }
+
+  constructor(
+    http: HttpClient,
+    config: AppConfigService,
+    @Inject(MOCK_CLASS_BUILDER) private classBuilder: (object: any) => T
```

Comment:
Why does this need to be injected? Why is it a 'mock' builder?

Reply:
`MOCK_CLASS_BUILDER` is used by the model.service unit tests. The reason it is a mock is because in the actual application `ModelService` is never actually used, only extended from. Any non service inputs in a service must have an injector so that the app.module file can instantiate default values. Thus to test the file, i need a mock injector which is only ever used by the model service unit tests.

Reply:
See [comment](https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r368373182) on naming. What about "stub" or "abstract"?